src/Controller/CRUDController.php

```diff
@@ -1478,6 +1483,48 @@ final protected function checkParentChildAssociation(Request $request, object $o
         }
     }
 
+    private function canBatchActionBeExecuted(string $action): bool
+    {
+        try {
+            return false !== $this->container
+                    ->get('controller_resolver')
+                    ->getController(new Request([], [], ['_controller' => $this->getBatchActionController($action)]));
+        } catch (\Throwable $error) {
+            return false;
+        }
+    }
+
+    /**
+     * @return string|callable|array<string, string> All must match symfony supported format for controller
+     */
+    private function getBatchActionController(string $action)
+    {
+        $batchActions = $this->admin->getBatchActions();
+        if (!\array_key_exists($action, $batchActions)) {
+            throw new \RuntimeException(sprintf('The `%s` batch action is not defined', $action));
+        }
+
+        $controller = $batchActions[$action]['controller'] ?? sprintf(
+            '%s::%s',
+            static::class,
+            sprintf('batchAction%s', InflectorFactory::create()->build()->classify($action))
+        );
+
+        return $controller;
+    }
+
+    private function getBatchActionExecutable(string $action): callable
+    {
+        $controller = $this->getBatchActionController($action);
+
+        return function (ProxyQueryInterface $query, Request $request) use ($controller) {
+            $request->attributes->set('_controller', $controller);
+            $request->attributes->set('query', $query);
```

Comment:
I find this a bit tricky, with this we are storing the whole query in session to be able to pass it to a controller.

`batchAction` needs to be an action because it's called directly from the view, but in this case we are processing the data sent, creating the query and making some other checks, so to me it seems more like a service implementing an interface receiving explicitly `ProxyQueryInterface $query` and `Request $request` rather than a controller that is not going to be called "as a controller".

On the other hand, if this new controller was able to replace the whole `batchAction`, I think that would be a good move (reusing somehow the `batchAction` code and would be a step forward to split/remove the `CRUDController`).

Reply:
> I find this a bit tricky, with this we are storing the whole query in session to be able to pass it to a controller.
> 
> `batchAction` needs to be an action because it's called directly from the view, but in this case we are processing the data sent, creating the query and making some other checks, so to me it seems more like a service implementing an interface receiving explicitly `ProxyQueryInterface $query` and `Request $request` rather than a controller that is not going to be called "as a controller".

Yeah, that's why I first proposed to use a service like this: https://github.com/sonata-project/SonataAdminBundle/pull/7819#discussion_r873200786
but it was considered as too complex. Do you have a better solution ?

> On the other hand, if this new controller was able to replace the whole `batchAction`, I think that would be a good move (reusing somehow the `batchAction` code and would be a step forward to split/remove the `CRUDController`).

How would you do this ?

Reply:
> I find this a bit tricky, with this we are storing the whole query in session to be able to pass it to a controller.

Not sure exactly what you mean by "session" the request attribute is not a session. Unless you didn't mean the php session.

It's just a simple solution to injection the parameter in the controller since Symfony check if the parameter is available in the request.

It's use often in symfony core you just have to look for **attributes->set(** in vendor/symfony.

Reply:
> rather than a controller that is not going to be called "as a controller".

Not sure I agree, a controller jobs is to take request validate the data and return the proper response. Which is exactly what it does.

From my point of view I consider that if the code don't need to be reuse (could also include automation test purpose) it doesn't need to be in a dedicated service.

In Symfony you can bind a service method to a route or inject a controller in another service/controller, so having it define as a service or controller is not that crucial. Rules of thumbs, the less amount lines (of code, documentation, configuration) needed to do something is (almost) always the best way. At the end it need to be as easy as possible for the end user to implement.

If you want to remove the CRUDController at some point we could move the code the batchAction to a single file controller/action.

Reply:
> From my point of view I consider that if the code don't need to be reuse (could also include automation test purpose) it doesn't need to be in a dedicated service.

If the same batch actions is used for differents admins, you'll reuse your batch action ; so the code is kinda reuse then.

Reply:
Sorry about the attributes an session, I got confused, what I meant is that this parameter bag I think is usually used for event listeners or services to add/retrieve custom information, but in this case it would be the user the one to have to call `$request->attributes->get('query')` which I don't find it simple or intuitive.

That's why I prefer explicit parameters like @VincentLanglet said, which I find it cleaner, but I can't think of right know of a good solution to register/fetch these actions easily by admin.

Maybe the ParamConverter/ArgumentResolver could be a solution so the user only has to define the action like:

```php
public function batchClone(BatchAction $batchAction): Response
{
    $admin = $batchAction->getAdmin();
    $query = $batchAction->getQuery();
    // whatever is needed
}
```

Reply:
> If the same batch actions is used for differents admins, you'll reuse your batch action ; so the code is kinda reuse then

Reuse in the sens of called from different flow, meaning "copy pasting" code causing duplication. This is not the case.

Reply:
> call `$request->attributes->get('query')` which I don't find it simple or intuitive.

Not really since by putting it in the attributes it will get injected in the controller method like I did put in my example. No need to manipulate anything.

> Maybe the ParamConverter/ArgumentResolver could be a solution so the user only has to define the action like:

I would suggest spiting this.

public function batchClone(ProxyQueryInterface $query, AdminInterface $admin): Response
{
    // whatever is needed
}

The $query is already done the way I have implement it, doing a param converter would not add anything.
The $admin could be done separately via a ParameterConverter that use the AdminFectcherInterface, that way we could reuse it every where (improving what we have to get the Admin)

Reply:
Just to give a real example of what it would look like after what I suggest for a real use case that I have for one of my client that rely on sonata media bundle that now have a final controller and a final Media admin:

```PHP
class MediaExtension extends AbstractAdminExtension
{
    public function configureBatchActions(AdminInterface $admin, array $actions): array
    {
        if ($admin instanceof BaseMediaAdmin) {
            $actions['allow_compression'] = [
                'label' => 'Allow compression',
                'ask_confirmation' => false,
                'controller' => MediaController::class . '::batchAllowCompressionAction',
            ];
        }
        
        return $actions;
    }
}

class MediaController extends AbstractController
{
    public function batchAllowCompressionAction(
        EntityManagerInterface $entityManager,
        ProxyQueryInterface $query,
        AdminInterface $admin
    ): Response {
        $medias = $query->execute();
        $compressionRequested = 0;

        try {
            /** @var Media $media */
            foreach ($medias as $media) {
                if ($media->setAllowCompression(true)) {
                    ++$compressionRequested;
                }
            }

            if ($compressionRequested) {
                $entityManager->flush();
            }

            $this->addFlash(
                'sonata_flash_success',
                sprintf('Media proceed - %s, compression requested - %s.', \count($medias), $compressionRequested)
            );
        } catch (Throwable $e) {
            $this->addFlash('sonata_flash_error', 'An error occur when allowing compression: ' . $e->getMessage());
        } finally {
            return new RedirectResponse(
                $admin->generateUrl('list', ['filter' => $admin->getFilterParameters()])
            );
        }
    }
}
```